lightning/src/util/persist.rs

```diff
@@ -406,30 +406,62 @@ where
 		)?;
 		let mut res = Vec::with_capacity(monitor_list.len());
 		// for each monitor...
-		for maybe_monitor_name in monitor_list.into_iter().map(MonitorName::new) {
-			let monitor_name = maybe_monitor_name?;
-			// ...parse the monitor
-			let (block_hash, monitor) = self.read_monitor(&monitor_name)?;
-			// ...parse and apply the updates with an id higher than the monitor.
-			let mut current_update_id = monitor.get_latest_update_id();
-			loop {
-				current_update_id = match current_update_id.checked_add(1) {
-					Some(next_update_id) => next_update_id,
-					None => break,
-				};
-				let update_name = UpdateName::from(current_update_id);
-				let update = match self.read_monitor_update(&monitor_name, &update_name) {
-					Ok(update) => update,
-					Err(err) if err.kind() == io::ErrorKind::NotFound => {
-						// We can't find any more updates, so we are done.
-						break;
-					}
-					Err(err) => return Err(err),
-				};
+		for monitor_key in monitor_list {
+			// ...push the result into the return vec
+			res.push(self.read_channel_monitor_with_updates(
+				broadcaster,
+				&fee_estimator,
+				monitor_key,
+			)?)
+		}
+		Ok(res)
+	}
 
-				monitor
-					.update_monitor(&update, broadcaster, fee_estimator.clone(), &self.logger)
-					.map_err(|e| {
+	/// Read a single channel monitor, along with any stored updates for them.
+	///
+	/// # EXTREMELY IMPORTANT
+	/// It is extremely important that your [`KVStore::read`] implementation uses the
+	/// [`io::ErrorKind::NotFound`] variant correctly: that is, when a file is not found, and _only_
+	/// in that circumstance (not when there is really a permissions error, for example). This is
+	/// because this function does not list updates. Instead, it lists all monitors, and then using
+	/// their stored `update_id`, synthesizes update storage keys, and tries them in sequence until
+	/// one is not found. All _other_ errors will be bubbled up in this function's [`io::Result`].
+	///
+	/// # Parallelizing
+	/// Loading a large number of monitors will be faster if done in parallel. You can use this
+	/// function to accomplish this. Take care to limit the number of parallel readers.
+	pub fn read_channel_monitor_with_updates<B: Deref, F: Deref + Clone>(
+		&self, broadcaster: &B, fee_estimator: &F, monitor_key: String,
+	) -> io::Result<(BlockHash, ChannelMonitor<<SP::Target as SignerProvider>::Signer>)>
+	where
+		ES::Target: EntropySource + Sized,
+		SP::Target: SignerProvider + Sized,
+		B::Target: BroadcasterInterface,
+		F::Target: FeeEstimator,
+	{
+		let monitor_name = MonitorName::new(monitor_key)?;
+		// ...parse the monitor
+		let (block_hash, monitor) = self.read_monitor(&monitor_name)?;
+		// ...parse and apply the updates with an id higher than the monitor.
```

Comment:
Can make this one comment, IMO.

Reply:
Just got rid of it. Ought to be pretty clear now.